File: stack/155-Min-Stack.py
class MinStack():

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val):
        self.stack.append(val)
        if len(self.min_stack) == 0 or val < self.stack[self.min_stack[-1]]:
            self.min_stack.append(len(self.stack) - 1)
            

    def pop(self):
        self.stack.pop()
        if len(self.min_stack) > 0 and self.min_stack[-1] == len(self.stack):
            self.min_stack.pop()
        

    def top(self):
        return self.stack[-1]
        

    def getMin(self):
        return self.stack[self.min_stack[-1]]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: stack/155-Min-Stack.go
type MinStack struct {
    mainStack []int
    minStack []int
}


func Constructor() MinStack {
    return MinStack {
        mainStack: []int{},
        minStack: []int{},
    }
}


func (this *MinStack) Push(val int)  {
    this.mainStack = append(this.mainStack, val)
    if len(this.minStack) == 0 {
        this.minStack = append(this.minStack, len(this.mainStack) - 1)
    } else {
        topIndexMin := this.minStack[len(this.minStack) - 1]
        if this.mainStack[topIndexMin] > val {
            this.minStack = append(this.minStack, len(this.mainStack) - 1)
        }
    }
}


func (this *MinStack) Pop()  {
    if len(this.mainStack) > 0 && len(this.minStack) > 0 {
        if this.minStack[len(this.minStack) - 1] == len(this.mainStack) - 1 {
            this.minStack = this.minStack[:len(this.minStack) - 1]
            this.mainStack = this.mainStack[:len(this.mainStack) - 1]
        } else {
            this.mainStack = this.mainStack[:len(this.mainStack) - 1]
        }
    }    
}


func (this *MinStack) Top() int {
    if len(this.mainStack) > 0 {
        return this.mainStack[len(this.mainStack) - 1]
    }
    return -1
}


func (this *MinStack) GetMin() int {
    index := this.minStack[len(this.minStack) - 1]
    return this.mainStack[index]
}


/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(val);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */